### PR TITLE
Optimize t-digest structure for small distributions

### PR DESCRIPTION
**Pull request by gastonar**
Dynamically increase the size of underlying arrays as values are added to digest.

### CODE REVIEW

**Comment by gastonar** [comment]
Instead of initializing 5 arrays of doubles of size 230, we dynamically increase the size of the arrays as we add more values. This helps save a lot of space in memory, since an empty/very small T-Digest can take up a lot of space. This change reduces the space in memory of an empty T-Digest from about 16_000 bytes to 300 bytes.

**Comment by gastonar** [comment]
Also fixes t-digest serialization documents, since it adds an extra field (`currentSize`) to the Slice when serializing.

File: presto-main/src/main/java/com/facebook/presto/tdigest/docs/t-digest.md [hunk 1]
```diff
@@ -21,6 +21,7 @@ The t-digest algorithm uses a set of centroids to store a distribution of values
 * compression: the compression factor used when creating the t-digest
 * total weight: the total number of values in the t-digest
 * activeCentroids: the number of centroids used by the t-digest
+* currentSize: the current size of the weight/mean arrays which represent the centroids
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/tdigest/docs/t-digest.md]: Could we use `activeCentroids` and not introduce `currentSize`.  I.e. isn' the active centroid count an effective proxy for the current size?

File: presto-main/src/main/java/com/facebook/presto/tdigest/TDigest.java [hunk 2]
```diff
@@ -108,21 +111,23 @@
...
+    private double[] mean;
 
     private double unmergedWeight;
 
     // this is the index of the next temporary centroid
     // this is a more Java-like convention than activeCentroids uses
     private int tempUsed;
-    private final double[] tempWeight;
-    private final double[] tempMean;
+    private final int size;
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/tdigest/TDigest.java]: This name is appropriate for a local variable but seems too terse for a member variable.  How about `maxCentroidCount` and `maxBufferSize`?

**Comment by gastonar** [comment]
All requested changes have been resolved. @tdcmeehan

**Comment by gastonar** [comment]
Benchmark results are directly on commit message. I'll add them here as well:

BenchmarkTDigest.benchmarkCopyBEFORE            N/A     avgt    30      3328.574 ±  79.584 ns/op
    BenchmarkTDigest.benchmarkCopyAFTER             N/A     avgt    30      1969.757 ±  19.348 ns/op
    BenchmarkTDigest.benchmarkDeserializeBEFORE     N/A     avgt    30      1623.594 ± 111.647 ns/op
    BenchmarkTDigest.benchmarkDeserializeAFTER      N/A     avgt    30       165.251 ±   5.514 ns/op
    BenchmarkTDigest.benchmarkInsertsBEFORE         N/A     avgt    30        67.205 ±   0.796 ns/op
    BenchmarkTDigest.benchmarkInsertsAFTER          N/A     avgt    30        69.787 ±   2.767 ns/op
    BenchmarkTDigest.benchmarkMergeBEFORE           N/A     avgt    30      4736.826 ±  24.052 ns/op
    BenchmarkTDigest.benchmarkMergeAFTER            N/A     avgt    30      2181.707 ±  49.311 ns/op
    BenchmarkTDigest.benchmarkSerializeBEFORE       N/A     avgt    30      2423.662 ± 138.643 ns/op
    BenchmarkTDigest.benchmarkSerializeAFTER        N/A     avgt    30       133.829 ±   4.896 ns/op

**Comment by tdcmeehan** [comment]
Replaced by #14020

**Pull request closed without merging**